spec/milestone_checker_spec.rb

```diff
@@ -0,0 +1,151 @@
+          },
+          'state' => 'open'
+        }
+
+        date_one_day_after_due = DateTime.parse('2023-07-01T23:00:01Z')
+
+        allow(@plugin.github).to receive(:pr_json).and_return(pr_json)
+        allow(DateTime).to receive(:now).and_return(date_one_day_after_due)
+
+        @plugin.check_milestone_due_date
+
+        expected_warning = ["This PR is assigned to the milestone [Release Day](https://wp.com) which has already expired.\nPlease make sure to get it merged by then or assign it to a later expiring milestone."]
+        expect(@dangerfile.status_report[:warnings]).to eq expected_warning
+      end
+
+      it 'reports a warning when a PR has a milestone with due date within a custom warning days threshold' do
+        pr_json = {
+          'milestone' => {
+            'title' => 'Release Day',
+            'html_url' => 'https://wp.com',
+            'due_on' => '2023-06-30T23:59:01Z'
+          },
+          'state' => 'open'
+        }
+
+        date_nine_days_before_due = DateTime.parse('2023-06-21T23:59:01Z')
+
+        allow(@plugin.github).to receive(:pr_json).and_return(pr_json)
+        allow(DateTime).to receive(:now).and_return(date_nine_days_before_due)
+
+        @plugin.check_milestone_due_date(warning_days: 10)
+
+        expected_warning = ["This PR is assigned to the milestone [Release Day](https://wp.com) which is expiring in less than 10 days.\nPlease make sure to get it merged by then or assign it to a later expiring milestone."]
+        expect(@dangerfile.status_report[:warnings]).to eq expected_warning
+      end
+
+      it 'does nothing when a PR has a milestone without a due date' do
+        pr_json = {
+          'milestone' => {
+            'title' => 'Release Day',
+            'html_url' => 'https://wp.com'
+          },
+          'state' => 'open'
+        }
+
+        allow(@plugin.github).to receive(:pr_json).and_return(pr_json)
+
+        @plugin.check_milestone_due_date
+
+        expect(@dangerfile.status_report[:warnings]).to be_empty
+      end
+
+      it 'does nothing when a PR has a milestone but already closed' do
+        pr_json = {
+          'milestone' => {
+            'title' => 'Release Day',
+            'html_url' => 'https://wp.com',
+            'due_on' => '2023-06-30T23:59:01Z'
+          },
+          'state' => 'closed'
+        }
+
+        allow(@plugin.github).to receive(:pr_json).and_return(pr_json)
+
+        @plugin.check_milestone_due_date
+
+        expect(@dangerfile.status_report[:warnings]).to be_empty
+      end
+
+      it 'warns when a PR without a milestone' do
+        allow(@plugin.github).to receive(:pr_json).and_return({})
+
+        @plugin.check_milestone_due_date
+
+        expected_warning = ['PR is not assigned to a milestone.']
+        expect(@dangerfile.status_report[:warnings]).to eq expected_warning
+      end
+
+      it 'does nothing when a PR does not have a milestone but this isnt required' do
```

Comment:
Nitpick

```suggestion
      it "does nothing when a PR does not have a milestone but this isn't required" do
```

---

Having said that, is this a parameter and behavior we need? 

I would assume that if we were to setup a repo that doesn't check for milestones, e.g. any of the library repos, we'd merely not add the check to the `Dangerfile`.

Reply:
I've re-read the diff and I think I get the intention: Use the same method to 1) check PRs have a milestone 2) check the milestone, if any, is still open.

Personally, I see those as two different checks, which could share implementation details under the hood.

- Ensure the PR has a milestone (similar to ensure the PR has a label)
- Ensure the milestone is still open (which implicitly assumes the PR should have a milestone and will fail/warn otherwise)

Keen to hear what you think.

Reply:
> I've re-read the diff and I think I get the intention: Use the same method to 1) check PRs have a milestone 2) check the milestone, if any, is still open.

Exactly; this is something inherited from the original Peril rule, just without the optionality on the milestone existence check.
I do agree it can be split into two separate checks, but now I'm not sure if the second check should also warn if the PR doesn't have a milestone. We could remove the check from the second, so that the caller can combine both calls if they want, having the freedom to have the milestone check or not (that was what I initially intended with the parameter):
```ruby
# Dangerfile

# will check if there's a milestone
milestone_checker.check_milestone_set

# will check if a milestone is close do its due date, but just return when there's no milestone set
milestone_checker.check_milestone_due_date
```

Having both calls is maybe what many repos will want, and we can have this behaviour in a common `Dangerfile` together with other common checks, so that other repos can just import it.

What do you think?

Reply:
> We could remove the check from the second, so that the caller can combine both calls if they want, having the freedom to have the milestone check or not (that was what I initially intended with the parameter)

🤔 

So far, my mental model has been "if you check that the milestone has a due date, then there should be an error if the milestone is not there". But I can see how this would work in a conditional mode such as "if there's a milestone, verify its due date is within the accepted range".

🤔 

From that angle, the current implementation gives more latitude.

What do you think of putting it all together, like this

```ruby
# Dangerfile

# will check if there's a milestone
milestone_checker.check_milestone_set

# will check if a milestone is close do its due date, but just return when there's no milestone set
milestone_checker.check_milestone_due_date

# will check if a milestone is close do its due date, and report an error if there's no milestone set
milestone_checker.check_milestone_due_date(if_no_milestone: :error)

# will check if a milestone is close do its due date, and report a warning if there's no milestone set
milestone_checker.check_milestone_due_date(if_no_milestone: :warn)
```

Reply:
I like the idea of the parameter like that. It makes the intention clearer plus providing the option to have warning/error/nothing when there's no milestone. I've pushed:
- Breaking down into two methods: 6d6a8e261222134cf7dfbd184eae67c698c774a7
- Wording update: 31f6aafa049ef8d6b1b8b557d8815165a28cf1db
- The new parameter: fbe857c7efa47c1a431da4084591a5941e6a96db